### PR TITLE
wxGUI: Integrate Grapical Modeler into single window layout

### PR DESCRIPTION
**Pull request by landam**
[Screencast from 2023-11-08 12-05-31.webm](https://github.com/OSGeo/grass/assets/5683186/505a603c-6640-411b-a520-531d44fa7c68)

Implemeneted:

- graphical modeler opens in main notebook when single layout enabled
- dock/undock icon added into toolbar
- title set correctly on model changed
- modeler settings replaced by new icon to avoid confusion with GUI settings icon
modeler-settings
- new icon added into toolbar to allow setting model properties (this functionality was available only from the menu)
options

Not implemented (will be solved in subsequent PRs):

- menu not shown (could be integrated in the main menu as a new "Modeler" item) 

Related changes in API:

- gmodeler code refactored (new modules `canvas` and `panels`)
- `display/map` like classes/methods/variables renamed (eg. `MapNotebook` renamed to `MainNotebook`, ...)

### CODE REVIEW

File: gui/wxpython/main_window/page.py
Comment (wenzeslaus) [inline gui/wxpython/main_window/page.py:107]: I think the disable is justified in this case, but the inline solution does not look right given there is a couple of those. Hard to say what is the correct solution for disabling here. Perhaps add that class to:

```
gui/wxpython/.pylintrc:322:ignored-classes=optparse.Values,thread._local,_thread._local
```

http://pylint-messages.wikidot.com/messages:e1101
https://pylint.pycqa.org/en/latest/user_guide/configuration/all-options.html#ignored-classes

Comment (landam) [reply gui/wxpython/main_window/page.py]: Changed in [f49a473](https://github.com/OSGeo/grass/pull/3003/commits/f49a47379da5af03ad08467375a9f82a4d9d74aa)

Comment (wenzeslaus) [reply gui/wxpython/main_window/page.py]: Thanks. I would say, code looks better, so let's see how this works out in the future.

**Review by lindakladivova — DISMISSED** [review]
Hi @landam, docking/undocking works great. I mainly like the MainPageBase class, the code is much more consistent and clear.

I just noticed small bugs in saving to *.gxm.

1) Modeler is docked - Save the current model to File - there is an issue when Ctrl+S triggers saving workspaces instead of the model file, the same problem for loading

2) Modeler is undocked - toolbar buttons work but Ctrl+S and Ctrl+O do nothing

3) Modeler is being undocked/docked- it automatically marks the modeler tab as unsaved (adds the asterisk to the title) although any changes have been previously made.

I also tested launching the r.what module via the modeler and when querying the coordinates in the module dialog, this bug appears: 

Traceback (most recent call last):
  File "/home/linduska/grass/dist.x86_64-pc-linux-
gnu/gui/wxpython/gui_core/gselect.py", line 2849, in
_onClick

self.mapWin = self._giface.GetMapWindow()
AttributeError
:
'GraphicalModelerGrassInterface' object has no attribute
'GetMapWindow'

**Comment by landam** [comment]
> 3. Modeler is being undocked/docked- it automatically marks the modeler tab as unsaved (adds the asterisk to the title) although any changes have been previously made.

@lindakladivova Window size is stored in model file. This is reason why ModelChanged() is called in OnSize(). In the case of dockable window it doesn't make sense so much.  Changed in d816fdd7e3

**Comment by landam** [comment]
> I also tested launching the r.what module via the modeler and when querying the coordinates in the module dialog, this bug appears:
> 
> Traceback (most recent call last): File "/home/linduska/grass/dist.x86_64-pc-linux- gnu/gui/wxpython/gui_core/gselect.py", line 2849, in _onClick
> 
> self.mapWin = self._giface.GetMapWindow() AttributeError : 'GraphicalModelerGrassInterface' object has no attribute 'GetMapWindow'

I was able to reproduce this bug in main branch. Reported as https://github.com/OSGeo/grass/issues/3470. So it doesn't seems to be related to this PR.

**Comment by landam** [comment]
> 1. Modeler is docked - Save the current model to File - there is an issue when Ctrl+S triggers saving workspaces instead of the model file, the same problem for loading
> 2. Modeler is undocked - toolbar buttons work but Ctrl+S and Ctrl+O do nothing

@lindakladivova Right, shortcuts simply don't work because dockable window has no menu.  This could be changed in future. 

For now shortcuts are not displayed when window is dockable, see 33e8219368a777da4f12059a18e2f49ceaf58e1a.

Graphical Modeler standalone:

image

Graphical Modeler dockable:

image

**Comment by landam** [comment]
@petrasovaa Please could you take a look at this PR? It would be nice to have this merged. Changes in `main_window` module should simplify rewriting other GUI components like eg. Python editor to be dockable.

**Comment by landam** [comment]
After merging this PR a new feature request must be reported: "show menu also for dockable components". Missing menu is not critical for Graphical Modeler (functionality available from the menu is also reachable from toolbar). But missing menu may be crucial for other components (like Python editor) when modifying them to be dockable.

**Comment by landam** [comment]
I added initial menu implementation in  eb918834b1

[Screencast_03_03_2024_03:29:45_PM.webm](https://github.com/OSGeo/grass/assets/5683186/534b3f85-7de6-4d63-880c-ab78787c73a9)

There are still some issues to be solved (like duplicated shortcuts). I am turning this PR back to draft.

**Comment by landam** [comment]
Tasks to be solved:

- [ ] duplicated shortcuts
- [ ] duplicated menu items when multiple modelers are open

**Comment by landam** [comment]
> @petrasovaa Please could you take a look at this PR? It would be nice to have this merged. Changes in `main_window` module should simplify rewriting other GUI components like eg. Python editor to be dockable.

Please wait with review when menu functionality will be finished.

**Comment by landam** [comment]
> @lindakladivova Right, shortcuts simply don't work because dockable window has no menu. This could be changed in future.

@lindakladivova Due to menu implementation I changed shortcuts to be unique in the whole menubar: ccc39f8cf7

Eg.

`Ctrl+N` is now `Ctrl+Alt+N`

The shortcuts are the same also for standalone window (`g.gui.gmodeler`). We could have different shortcuts for dockable version (`Ctrl+Alt+O`) and standalone version (`Ctrl+O`). But it would be probably confusing for the user.

[Screencast_03_03_2024_05:56:27_PM.webm](https://github.com/OSGeo/grass/assets/5683186/785ad0bf-3b4e-4403-90cb-d390e6bccf03)

**Comment by landam** [comment]
Bug to be solved:

0. Open modeler
1. Undock modeler
2. Dock modeler
3. Go to menu `File`

```py
Traceback (most recent call last):
  File "/home/martin/src/grass/dist.x86_64-pc-linux-
gnu/gui/wxpython/gui_core/menu.py", line 145, in
OnMenuHighlight

item = self.FindItemById(id)
RuntimeError
:
wrapped C/C++ object of type Menu has been deleted
```

**Comment by landam** [comment]
> * [x]  Bug to be solved ([33b1119]

I am still not sure why already destroyed menu objects are causing such problem. I got around this problem by reusing the already created menu object, see 33b1119. See also screencast below for demostration

[Screencast_03_03_2024_09:51:30_PM.webm](https://github.com/OSGeo/grass/assets/5683186/c2f1c97b-5576-484f-a958-6288171b94b8)

**Comment by landam** [comment]
> > @petrasovaa Please could you take a look at this PR? It would be nice to have this merged. Changes in `main_window` module should simplify rewriting other GUI components like eg. Python editor to be dockable.
> 
> Please wait with review until the menu functionality will be finished.

All issues were addressed, PR is ready to review.

File: gui/wxpython/mapdisp/frame.py
Comment (petrasovaa) [inline gui/wxpython/mapdisp/frame.py:127]: It looks like this disappeared and it's causing an error when saving image in map display:
```
Traceback (most recent call last):
  File "/home/anna/dev/grass/grass/dist.x86_64-pc-linux-
gnu/gui/wxpython/mapdisp/frame.py", line 729, in SaveToFile

dlg = ImageSizeDialog(self,
img_size=self._saved_output_img_size)
AttributeError
:
'MapPanel' object has no attribute '_saved_output_img_size'
```

File: gui/wxpython/gmodeler/toolbars.py
Comment (petrasovaa) [inline gui/wxpython/gmodeler/toolbars.py:189]: This shows "Un(dock) Map Display"

**Review by petrasovaa — CHANGES_REQUESTED** [review]

File: gui/wxpython/gmodeler/toolbars.py
Comment (landam) [reply gui/wxpython/gmodeler/toolbars.py]: Good catch, fixed in 83a4a4c753

File: gui/wxpython/mapdisp/frame.py
Comment (landam) [reply gui/wxpython/mapdisp/frame.py:127]: Thanks for testing. Fixed in 78ad9a7be2.

**Review by petrasovaa — APPROVED** [review]
I will approve this now, but please address the new bandit security warning, the one about eval. It looks like ignoring it by adding `# nosec B602` on that line should work.

**Review by petrasovaa — DISMISSED** [review]

**Review by pesekon2 — DISMISSED** [review]

**Comment by landam** [comment]
> and there are more probably... It would be a good idea to get rid of the star imports and test flake8 on this.

Addressed by [cd78874](https://github.com/OSGeo/grass/pull/3003/commits/cd788741f54360554a96a5a3e3867ff000f3704b)

**Comment by petrasovaa** [comment]
> > and there are more probably... It would be a good idea to get rid of the star imports and test flake8 on this.
> 
> Addressed by [cd78874](https://github.com/OSGeo/grass/pull/3003/commits/cd788741f54360554a96a5a3e3867ff000f3704b)

Thanks, it would be good idea to remove the ignored errors from the .flake8 config file.

**Comment by landam** [comment]
> Thanks, it would be good idea to remove the ignored errors from the .flake8 config file.

Yes, done in c1c010f

**Comment by pesekon2** [comment]
When I open the modeler in the single-window layout, I have some issue with the gmodeler icons. They seem invisible until I drag the mouse over them (see the attached screencast). When I click on another window, all of them appear.

[Screencast from 6.4.2024 13:59:55.webm](https://github.com/OSGeo/grass/assets/17025405/a963a044-e25e-43d2-bdc4-473551beb0f0)

@landam  @petrasovaa: Does it happen to you in this PR too? It might be just my local issue so if you don't encounter this problem, I will not block the PR and try to find the solution myself later.

**Comment by landam** [comment]
> @landam @petrasovaa: Does it happen to you in this PR too? It might be just my local issue so if you don't encounter this problem, I will not block the PR and try to find the solution myself later.

Thanks for testing. I am not aware of this issue, it never happened to me. See screencast below.

[Screencast from 2024-04-06 20-33-29.webm](https://github.com/OSGeo/grass/assets/5683186/1a32b76a-dbb0-4857-a255-e4a0f546e022)

In your screencast I also don't see any tabs. So it seems that also other widgets from the gmodeler panel have the same issue.

@pesekon2 Which wxPython version do you have?

**Comment by pesekon2** [comment]
> In your screencast I also don't see any tabs. So it seems that also other widgets from the gmodeler panel have the same issue.

You are right. I see the same behaviour in tabs. They all appear at once once I go there with the mouse there.

> Which wxPython version do you have?

4.2.1. It should be the newest version. Python 3.10.12.

**Comment by landam** [comment]
> 4.2.1. It should be the newest version. Python 3.10.12.

No clue. 

Python: 3.11.8
wxPython: 4.2.1

**Comment by pesekon2** [comment]
> No clue.
> 
> Python: 3.11.8 wxPython: 4.2.1

Then if nobody else encounters the same issue, I will assume it is a problem at my side. I will not block this PR then (but cannot approve either, as I don't see the same weird behaviour in the old-school gmodeler).

I didn't find any other issue with the PR.

**Comment by landam** [comment]
> Python: 3.11.8 wxPython: 4.2.1

I also tested this PR on Python 3.10.12 (with wxPython 4.2.1). No problem detected.

**Review by petrasovaa — APPROVED** [review]

**Pull request merged**
